drivers/bus/ancillary_bus.c

```diff
@@ -0,0 +1,213 @@
+{
+	struct ancillary_driver *adrv = to_ancillary_drv(drv);
+	struct ancillary_device *adev = to_ancillary_dev(dev);
+
+	return !!ancillary_match_id(adrv->id_table, adev);
+}
+
+static int ancillary_uevent(struct device *dev, struct kobj_uevent_env *env)
+{
+	struct ancillary_device *adev = to_ancillary_dev(dev);
+
+	if (add_uevent_var(env, "MODALIAS=%s%s", ANCILLARY_MODULE_PREFIX,
+			   adev->match_name))
+		return -ENOMEM;
+
+	return 0;
+}
+
+static const struct dev_pm_ops ancillary_dev_pm_ops = {
+	SET_RUNTIME_PM_OPS(pm_generic_runtime_suspend,
+			   pm_generic_runtime_resume, NULL)
+	SET_SYSTEM_SLEEP_PM_OPS(pm_generic_suspend, pm_generic_resume)
+};
+
+struct bus_type ancillary_bus_type = {
+	.name = "ancillary",
+	.match = ancillary_match,
+	.uevent = ancillary_uevent,
+	.pm = &ancillary_dev_pm_ops,
+};
+
+/**
+ * ancillary_release_device - Destroy a ancillary device
+ * @_dev: device to release
+ */
+static void ancillary_release_device(struct device *dev)
+{
+	struct ancillary_device *adev = to_ancillary_dev(dev);
+	u32 ida = adev->id;
+
+	adev->release(adev);
+	if (ida != ANCILLARY_INVALID_ID)
+		ida_simple_remove(&ancillary_dev_ida, ida);
+}
+
+/**
+ * ancillary_register_device - add a ancillary bus device
+ * @adev: ancillary bus device to add
+ */
+int ancillary_register_device(struct ancillary_device *adev)
+{
+	int ret;
+
+	if (!adev->release) {
+		dev_err(&adev->dev, "release callback not set for adev!\n");
+		return -EINVAL;
+	}
+
+	/* All error paths out of this function after the device_initialize
+	 * must perform a put_device() so that the .release() callback is
+	 * called for an error condition.
+	 */
+	device_initialize(&adev->dev);
+
+	adev->dev.bus = &ancillary_bus_type;
+	adev->dev.release = ancillary_release_device;
+
+	/* All device IDs are automatically allocated */
+	ret = ida_simple_get(&ancillary_dev_ida, 0, 0, GFP_KERNEL);
+	if (ret < 0) {
+		adev->id = ANCILLARY_INVALID_ID;
+		dev_err(&adev->dev, "get IDA idx for ancillary device failed!\n");
+		goto err;
+	}
+
+	adev->id = ret;
+
+	ret = dev_set_name(&adev->dev, "%s.%d", adev->match_name, adev->id);
+	if (ret) {
+		dev_err(&adev->dev, "dev_set_name failed for device\n");
+		goto err;
+	}
+
+	dev_dbg(&adev->dev, "Registering ancillary device '%s'\n",
+		dev_name(&adev->dev));
+
+	ret = device_add(&adev->dev);
+	if (!ret)
+		return ret;
+
+	dev_err(&adev->dev, "Add device to ancillary device failed!: %d\n", ret);
+
+err:
+	put_device(&adev->dev);
+
+	return ret;
+}
+EXPORT_SYMBOL_GPL(ancillary_register_device);
+
+static int ancillary_probe_driver(struct device *_dev)
+{
+	struct ancillary_driver *adrv = to_ancillary_drv(_dev->driver);
+	struct ancillary_device *adev = to_ancillary_dev(_dev);
+	int ret;
+
+	ret = dev_pm_domain_attach(_dev, true);
+	if (ret) {
+		dev_warn(_dev, "Failed to attach to PM Domain : %d\n", ret);
+		return ret;
+	}
+
+	ret = adrv->probe(adev);
+	if (ret) {
+		dev_err(&adev->dev, "Probe returned error\n");
```

Comment:
Please be consistent, always use dev or adev->dev in doing dev_err()..

dev_err(dev) is preferred.

Reply:
changed